### PR TITLE
fix(rule): update avoided emissions processor and test cases for new baseline and subtype logic

### PR DESCRIPTION
**Pull request by AMarcosCastelo**
### Summary

Update the Avoided Emissions processor and related test cases to support new business logic for baselines and subtypes, including improved error handling and matrix-driven calculations.

### Details

- Refactored the avoided emissions processor to:
  - Use `MethodologyBaseline` and `MassIdOrganicSubtype` enums for all baseline logic.
  - Accept partial baseline objects for more flexible homologation document structures.
  - Use a matrix (`AVOIDED_EMISSIONS_BASELINE_MATRIX`) for baseline values, supporting all organic subtypes and baseline types.
  - Update the calculation to: `(1 - exceedingEmissionCoefficient) * baseline * massIdDocumentValue`.
  - Improve error handling for missing/invalid baselines and subtypes.
- Updated all related test cases to:
  - Use the new enums and matrix structure.
  - Check for new error messages and calculation logic.
  - Add edge cases for missing/invalid data.
- Updated typings to allow partial baseline objects.
- Cleaned up and clarified test scenarios and processor logic.

---

- [x] **I, the PR author, declare that this PR works as expected and does not break any service. I also declare that I gave my best to apply all of the best practices and that I'm leaving the code better than I found it.**

## Summary by CodeRabbit

- **New Features**
  - Enhanced avoided emissions processing to incorporate waste generator baseline data, enabling more accurate baseline-dependent calculations.
  - Added support for new document types and attributes, including recycling baselines and exceeding emission coefficients.
  - Expanded error handling with specific messages for missing or invalid waste generator data.
- **Documentation**
  - Introduced comprehensive documentation for the baseline emission factors module, detailing methodology and calculation context.
- **Tests**
  - Extended and updated test cases to cover new scenarios, attributes, and error conditions related to baseline and subtype validation.

### CODE REVIEW

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# BOLD Carbon CH₄ – Baseline Emission Factors
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md]: I think this processor is much more than _Baseline Emission Factors_, I would go with name _Avoided Emission Calculation_ or something similar.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+## 📘 Context
+
+When calculating avoided emissions for composting projects, it is essential to compare each material's potential emissions under traditional waste disposal scenarios. These reference scenarios (baselines) are:
+
+- **Landfills with methane flaring**
+- **Landfills without methane flaring**
+- **Open-air dumps**
+
+Each organic waste category is assigned emission factors (in **tons of CO₂e per ton of waste**) based on default values and decay models aligned with **AMS-III.F v12.0** and IPCC standards.
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md]: Just to make sure, is *decay* the right word in this context? Is this based on some reference text?

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.errors.ts [hunk 3]
```diff
@@ -2,9 +2,15 @@ import { BaseProcessorErrors } from '@carrot-fndn/shared/methodologies/bold/proc
...
+    INVALID_MASS_ID_DOCUMENT_SUBTYPE:
+      'The "MassID" document has an invalid subtype.',
+    INVALID_WASTE_GENERATOR_BASELINES:
+      'The "Waste Generator homologation" document has no valid baselines.',
     MISSING_MASS_ID_DOCUMENT: 'The "MassID" document was not found.',
     MISSING_RECYCLER_HOMOLOGATION_DOCUMENT:
       'The "Recycler homologation" document was not found.',
+    MISSING_WASTE_GENERATOR_HOMOLOGATION_DOCUMENT:
+      'The "Waste Generator homologation" document was not found.',
     REJECTED_BY_ERROR: 'Unable to validate the avoided-emissions process.',
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.errors.ts:14]: [nitpicking] Considering that these messages are used in several rules, we could have them shared across them all to make sure we are always using the same messages.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.types.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+export interface RuleSubject {
+  exceedingEmissionCoefficient:
+    | MethodologyDocumentEventAttributeValue
+    | undefined;
+  massIdDocumentValue: number;
+  wasteGeneratorBaseline: MethodologyBaseline | undefined;
+  wasteSubtype: MassIdOrganicSubtype;
+}
+
+export type WasteGeneratorBaselinesValue = Partial<
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.types.ts]:
```suggestion
export type WasteGeneratorBaselineValues = Partial<
```

File: libs/shared/types/src/methodology/methodology-enum.types.ts [hunk 5]
```diff
@@ -62,6 +62,7 @@ export enum MethodologyDocumentEventName {
   OUTPUT = 'OUTPUT',
   PICK_UP = 'Pick-up',
   RECYCLED = 'Recycled',
+  RECYCLING_BASELINE = 'Recycling Baseline',
```
Comment (sangalli) [inline libs/shared/types/src/methodology/methodology-enum.types.ts:65]:
```suggestion
  RECYCLING_BASELINES = 'Recycling Baselines',
```

File: libs/shared/methodologies/bold/testing/src/builders/bold-participant-homologation.stubs.ts [hunk 6]
```diff
@@ -92,6 +100,29 @@ export const stubBoldEmissionAndCompostingMetricsEvent = ({
     ),
   );
 
+const defaultRecyclingBaselineEventMetadataAttributes: MetadataAttributeParameter[] =
```
Comment (sangalli) [inline libs/shared/methodologies/bold/testing/src/builders/bold-participant-homologation.stubs.ts]:
```suggestion
const defaultRecyclingBaselinesEventMetadataAttributes: MetadataAttributeParameter[] =
```

File: libs/shared/methodologies/bold/testing/src/builders/bold-participant-homologation.stubs.ts [hunk 7]
```diff
@@ -92,6 +100,29 @@ export const stubBoldEmissionAndCompostingMetricsEvent = ({
...
+  [
+    [
+      BASELINES,
+      { [random<MassIdOrganicSubtype>()]: random<MethodologyBaseline>() },
+    ],
+  ];
+
+export const stubBoldRecyclingBaselineEvent = ({
```
Comment (sangalli) [inline libs/shared/methodologies/bold/testing/src/builders/bold-participant-homologation.stubs.ts]:
```suggestion
export const stubBoldRecyclingBaselinesEvent = ({
```

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts [hunk 8]
```diff
@@ -32,29 +34,37 @@ import {
...
 export const RESULT_COMMENTS = {
   APPROVED: (
     avoidedEmissions: number,
-    emissionIndex: number,
+    exceedingemissionCoefficient: number,
     currentValue: number,
   ) =>
-    `The avoided emissions were calculated as ${avoidedEmissions} kg CO₂e using the formula ${emissionIndex} × ${currentValue} = ${avoidedEmissions} [formula: emission_index × current_value = avoided_emissions].`,
-  MISSING_EMISSION_FACTOR: `The "${EMISSION_FACTOR}" attribute was not found in the "Recycler Homologation" document or it is invalid.`,
+    `The avoided emissions were calculated as ${avoidedEmissions} kg CO₂e using the formula (1 - ${exceedingemissionCoefficient}) × ${currentValue} = ${avoidedEmissions} [formula: (1 - emission_index) × current_value = avoided_emissions].`,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts]: Is this formula actually correct? I feel like it is missing the _avoidedEmissionsByMaterialAndBaselinePerTon_.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import {
+  MassIdOrganicSubtype,
+  MethodologyBaseline,
+} from '@carrot-fndn/shared/methodologies/bold/types';
+
+export const AVOIDED_EMISSIONS_BASELINE_MATRIX: Record<
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]:
```suggestion
export const AVOIDED_EMISSIONS_BY_MATERIAL_AND_BASELINE_PER_TON: Record<
```

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts [hunk 10]
```diff
@@ -81,21 +91,42 @@ export class AvoidedEmissionsProcessor extends RuleDataProcessor {
...
+          RESULT_COMMENTS.MISSING_RECYCLING_BASELINE_FOR_WASTE_SUBTYPE(
+            wasteSubtype,
+          ),
         resultStatus: RuleOutputStatus.REJECTED,
       };
     }
 
-    const avoidedEmissions = emissionFactor * massIdDocumentValue;
+    const avoidedEmissionsBaseline =
+      AVOIDED_EMISSIONS_BASELINE_MATRIX[wasteSubtype][wasteGeneratorBaseline];
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts]: Don't we need a validation here to make sure that the return value is not `undefined`?

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts [hunk 11]
```diff
@@ -81,21 +91,42 @@ export class AvoidedEmissionsProcessor extends RuleDataProcessor {
...
+
+    const avoidedEmissions =
+      (1 - exceedingemissionCoefficient) *
+      avoidedEmissionsBaseline *
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts]: I know this is a big name, but I would like to be more clear about what this variable means, such as:
```suggestion
      avoidedEmissionsByMaterialAndBaselinePerTon *
```

Instead of `material`, we could also use `wasteSubtype`.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  [MassIdOrganicSubtype.EFB_SIMILAR_TO_GARDEN_YARD_AND_PARK_WASTE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.629,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 1.251,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.975,
+  },
+  [MassIdOrganicSubtype.FOOD_FOOD_WASTE_AND_BEVERAGES]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.459,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 0.94,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.727,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: If we use the spreadsheet with 1 ton, these factor is 0.726, but if we use 1000 tons it is 0.727, so there is some small imprecision here. I think we should round down to be conservative. Another approach is to use more decimal cases.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  [MassIdOrganicSubtype.GARDEN_YARD_AND_PARK_WASTE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.5,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 1.121,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.845,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: Here is the same rounding difference when using just one ton: 0.5 becomes 0.499 and 0.845 becomes 0..844.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+> = {
+  [MassIdOrganicSubtype.DOMESTIC_SLUDGE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.067,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 0.227,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.156,
+  },
+  [MassIdOrganicSubtype.EFB_SIMILAR_TO_GARDEN_YARD_AND_PARK_WASTE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.629,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 1.251,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.975,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: Here is the same rounding difference when using just one ton: 1.251 becomes 1.250 and 0.975 becomes 0..974.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.727,
+  },
+  [MassIdOrganicSubtype.GARDEN_YARD_AND_PARK_WASTE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.5,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 1.121,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.845,
+  },
+  [MassIdOrganicSubtype.INDUSTRIAL_SLUDGE]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.224,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 0.513,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: Here is the same rounding difference when using just one ton: 0.224 becomes 0.223 and 0.513 becomes 0..512.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.384,
+  },
+  [MassIdOrganicSubtype.TOBACCO]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.459,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 0.94,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 0.727,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: Review my comment in food waste regarding rounding here

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  },
+  [MassIdOrganicSubtype.WOOD_AND_WOOD_PRODUCTS]: {
+    [MethodologyBaseline.LANDFILLS_WITH_FLARING_OF_METHANE_GAS]: 0.72,
+    [MethodologyBaseline.LANDFILLS_WITHOUT_FLARING_OF_METHANE_GAS]: 1.416,
+    [MethodologyBaseline.OPEN_AIR_DUMP]: 1.107,
```
Comment (sangalli) [inline libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.constants.ts]: Here is the same rounding difference when using just one ton: 1.416 becomes 1.415 and 1.107 becomes 1.106.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md (hunk 2 above)
Comment (AMarcosCastelo) [reply libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md]: I asked to the AI to summarize the Bold Carbon Avoided Emissions using the BOLD Carbon methodology PDF. I will send this message to Laura and Stella to validate the information. What do you think?

[AMS-III.F v12.0 – UNFCCC Methodology](https://cdm.unfccc.int/methodologies/DB/NZ83KB7YHBIA7HL2U1PCNAOCHPUQYX)

[Carrot BOLD Carbon CH₄ Methodology PDF](BOLD Carbon CH4 v1.0.pdf)

[IPCC GHG Protocol (AR6)](https://www.ipcc.ch/report/ar6/)

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.errors.ts (hunk 3 above)
Comment (AMarcosCastelo) [reply libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.errors.ts:14]: Gooooood Point!! I will add this to the `methodology-rules` improvements and work on it later.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts (hunk 10 above)
Comment (AMarcosCastelo) [reply libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/avoided-emissions.processor.ts]: We validate it on another level. If it passes any waste subtype or a different baseline, it will throw an error before.

File: libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md (hunk 2 above)
Comment (sangalli) [reply libs/methodologies/bold/rule-processors/mass-id/avoided-emissions/src/README.md]: It makes sense to me!

File: libs/shared/methodologies/bold/types/src/enum.types.ts [hunk 18]
```diff
@@ -97,6 +98,8 @@ export enum DocumentEventName {
   OUTPUT = MethodologyDocumentEventName.OUTPUT,
   PICK_UP = MethodologyDocumentEventName.PICK_UP,
   RECYCLED = MethodologyDocumentEventName.RECYCLED,
+  RECYCLING_BASELINE = MethodologyDocumentEventName.RECYCLING_BASELINE,
```
Comment (sangalli) [inline libs/shared/methodologies/bold/types/src/enum.types.ts]: Where are we using this one?

**Review by GLGuilherme — DISMISSED** [review]

**Review by sangalli — APPROVED** [review]

**Review by andtankian — APPROVED** [review]
